cranelift/isle/isle/src/overlap.rs

```diff
@@ -0,0 +1,876 @@
+//! Overlap detection for rules in ISLE.
+
+use crate::error::{Error, Result, Source, Span};
+use crate::sema::{
+    self, ConstructorKind, Rule, RuleId, Sym, Term, TermEnv, TermId, TermKind, Type, TypeEnv,
+    TypeId, VarId,
+};
+
+/// Check for overlap.
+pub fn check(tyenv: &TypeEnv, termenv: &TermEnv) -> Result<()> {
+    let env = Env::new(tyenv, termenv);
+    let mut errors = Errors::new(termenv.rules.len());
+    for term in termenv.terms.iter() {
+        // The only isle declaration that currently produces overlap is constructors whose
+        // definition is entirely in isle.
+        if !env.is_internal_constructor(term.id) {
+            continue;
+        }
+
+        check_overlap_groups(&mut errors, &env, term);
+    }
+
+    if !errors.is_empty() {
+        let mut errors = errors.report(&env);
+        return match errors.len() {
+            1 => Err(errors.pop().unwrap()),
+            _ => Err(Error::Errors(std::mem::take(&mut errors))),
+        };
+    }
+
+    Ok(())
```

Comment:
I think `errors.report` will correctly return an empty vector if there are no errors, so I don't think you need the `is_empty` check up front. Also I think Rust will let you match directly on the returned vector:

```suggestion
    match errors.report(&env) {
        [] => Ok(()),
        [e] => Err(e),
        errors => Err(Error::Errors(errors)),
    }
```